Implement basic support for Objective-C methods.

jacob-carlborg opened the pull request:
Some return types does not work, those are: `long double`, `_Complex long double` and structs that are too large to fit in registers. The runtime part of this pull is in https://github.com/D-Programming-Language/druntime/pull/1111.

The below text is in the commit message as well. I don't know if it's too long for a commit message or if there's a better place to document this.

Basic support for classes, interfaces and instance methods.
This is implemented by adding a new linkage attribute, `Objective-C`,
and a compiler recognized UDA, `@selector`. The linkage attribute is
to be used on a class or interface. The UDA is attached to a method.

The linkage attribute tells the compiler that the class should use the
name mangling that matches the one used by Objective-C
(same as C, no mangling) and that all methods in the class
should use the Objective-C way of calling methods, see below.
The calling convention for Objective-C methods and functions is the
same as for C.

The selector UDA tells the compiler what Objective-C selector the
method should have. The selector is used in the Objective-C runtime
to find the implementation of a given method.

An Objective-C method call is implemented by making a regular C call
to the `objc_msgSend` function in the Objective-C runtime.
The signature of `objc_msgSend` looks something like this:

`id objc_msgSend(id self, SEL op, ...);`
- The first parameter is the object (this/self pointer)
- The second parameter is the selector attached to the method
- The last parameter is for all the arguments that the
  implementation expects

The call to `objc_msgSend` should not be performed as a variadic
call but instead as if it had the same signature as the method
that should be called but with the two additional parameter,
`self` and `op`, added first. The implementation of `objc_msgSend`
will jump to the method instead of calling it.

Because of the above, multiple versions exist of `objc_msgSend`.
Depending on the return type of the method that is called the correct
version need to be used. This depends on the ABI. This is a list of
functions and for which types they're used on OS X 64bit:
- objc_msgSend_stret - Used for structs too large to be returned in
  registries
- objc_msgSend_fpret - Used for `long double`
- objc_msgSend_fp2ret - Used for `_Complex long double`
- objc_msgSend - Used for everything else

michelf committed:
Implement basic support for Objective-C methods.

Basic support for classes, interfaces and instance methods.
This is implemented by adding a new linkage attribute, `Objective-C`,
and a compiler recognized UDA, `@selector`. The linkage attribute is
to be used on a class or interface. The UDA is attached to a method.

The linkage attribute tells the compiler that the class should use the
name mangling that matches the one used by Objective-C
(same as C, no mangling) and that all methods in the class
should use the Objective-C way of calling methods, see below.
The calling convention for Objective-C methods and functions is the
same as for C.

The selector UDA tells the compiler what Objective-C selector the
method should have. The selector is used in the Objective-C runtime
to find the implementation of a given method.

An Objective-C method call is implemented by making a regular C call
to the `objc_msgSend` function in the Objective-C runtime.
The signature of `objc_msgSend` looks something like this:

`id objc_msgSend(id self, SEL op, ...);`

* The first parameter is the object (this/self pointer)
* The second parameter is the selector attached to the method
* The last parameter is for all the arguments that the
implementation expects

The call to `objc_msgSend` should not be performed as a variadic
call but instead as if it had the same signature as the method
that should be called but with the two additional parameter,
`self` and `op`, added first. The implementation of `objc_msgSend`
will jump to the method instead of calling it.

Because of the above, multiple versions exist of `objc_msgSend`.
Depending on the return type of the method that is called the correct
version need to be used. This depends on the ABI. This is a list of
functions and for which types they're used on OS X 64bit:

* objc_msgSend_stret - Used for structs too large to be returned in
registries
* objc_msgSend_fpret - Used for `long double`
* objc_msgSend_fp2ret - Used for `_Complex long double`
* objc_msgSend - Used for everything else

jacob-carlborg committed:
Add tests for the various objc_msgSend_* functions.

ZombineDev commented on src/objc_symbols_common.c:
```diff
@@ -0,0 +1,104 @@
+    stringtable->_init();
+
+    return stringtable;
+}
```
You can remove this function and call [`StringTable::reset(..)`][1] instead, which does the same thing.

[1]: https://github.com/D-Programming-Language/dmd/blob/master/src/root/stringtable.c#L135

jacob-carlborg replied:
Then I don't have to new the string table?

ZombineDev replied:
Yes, AFAIU.

jacob-carlborg replied:
Ok, cool, I'll update that.

ZombineDev commented on src/objc_symbols_common.c:
```diff
@@ -0,0 +1,104 @@
+
+    // clear tables
+    smethvarnametable = initStringTable(smethvarnametable);
+    smethvarreftable = initStringTable(smethvarreftable);
```
You can do it like this:
```
smethvarnametable->reset();
smethvarreftable->reset();
```

ZombineDev commented on src/objc_stubs.c:
```diff
@@ -0,0 +1,205 @@
+{
+    assert(false && "Should never be called when D_OBJC is false");
+    return NULL;
+}
```
I think this stub (`initStringTable`) is also not needed.

ZombineDev commented on src/objc_func_selector.c:
```diff
@@ -0,0 +1,65 @@
+#include "objc.h"
+#include "scope.h"
+
+void objc_FuncDeclaration_semantic_setSelector(FuncDeclaration *self, Scope *sc)
```
You should add a stub for `objc_FuncDeclaration_semantic_setSelector(..)` in objc_stubs.h, because that way you can avoid including this file if D_OBJC != 1 in the posix.mak file.

jacob-carlborg replied:
This shouldn't be a stub, that's why it's a separate file so I can include regardless if `D_OBJC` is enabled or not.

ZombineDev replied:
Why do you want to enable it if `D_OBJC` is not enabled?

ZombineDev replied:
[`objc_FuncDeclaration_semantic_validateSelector`][1] is also a stub.

[1]: https://github.com/D-Programming-Language/dmd/pull/4321/files#diff-43282ebf5a2de5fdbcb3b5083ddf949dR1214

jacob-carlborg replied:
It's implemented [here](https://github.com/D-Programming-Language/dmd/pull/4321/files#diff-ea795c64edf6f31b4615d946696ac5f3R179), not as a stub. I can't think of a really good reason now but I think the original reasons why I did this were:

* To split the implementation of extracting the selector and validating the selector in two functions
* To have an extra check in the compiler if `core.attribute.selector` is defined even though `D_ObjectiveC` is not enabled

ZombineDev replied:
As far as i can [see][1] `objc_FuncDeclaration_semantic_validateSelector(..)` is a stub if `D_OBJC ` is enabled.

[1]: https://github.com/D-Programming-Language/dmd/pull/4321/files#diff-ea795c64edf6f31b4615d946696ac5f3R179

jacob-carlborg replied:
Yes. I'll stub `objc_FuncDeclaration_semantic_setSelector` and `objc_FuncDeclaration_semantic_validateSelector` instead.

ZombineDev commented:
I think the autotester is currently failing on Windows, because you need to include `objc_stubs.c` to `win32.mak`, similar to what you have done for `posix.mak`.

schuetzm commented on src/idgen.c:
```diff
@@ -353,6 +354,9 @@ Msgtable msgtable[] =
     { "char_traits" },
+
+    // Compiler recognized UDA's
+    { "udaSelector", "core.attribute.selector" },
```
Would it be better not to detect the special UDA by name, but have the user import e.g. `core.objc`, which contains a symbol `selector`, and let the compiler detect this special symbol? Then it wouldn't make existing `@selector` UDAs invalid, at the cost of needing an explicit import.

jacob-carlborg replied:
This is how it works.

jacob-carlborg commented:
> I think the autotester is currently failing on Windows, because you need to include objc_stubs.c to win32.mak, similar to what you have done for posix.mak.

Ok, I see. I'll add it to win32.mak.

ZombineDev commented on src/objc_func_selector.c:
```diff
@@ -0,0 +1,65 @@
+                    StringExp *se = (*literal->elements)[0]->toStringExp();
+                    assert(se);
+
+                    self->objc.selector = ObjcSelector::lookup((const char *)se->toUTF8(sc)->string);
```
`ObjcSelector::lookup(..)` is a stub, so you are just delaying the error, afaics.

jacob-carlborg replied:
Yes, that's a problem.

jacob-carlborg commented:
@ZombineDev a couple of updates:

* Moved `objc_FuncDeclaration_semantic_setSelector` to `objc_func.c` and stubbed it on platforms where `D_OBJC` is disabled

* Removed the stub for `initStringTable`
* Updated the Windows makefile

I couldn't use `stringtable->reset()` cause it's a member function and the original `initStringTable` creates the actual object.

@WalterBright  @yebblies @9rnsr please review.

yebblies commented:
There are some parts of this that don't match the usual dmd style - eg ref parameters, the style of asserts, calling parameters `self` instead of the usual `fd`.

There also seem to be a _lot_ of small new files.  Is this because later pull requests are going to flesh them out a lot more?

Is `ObjcSelectorBuilder` really necessary?  Can't you just use `OutBuffer`?

yebblies commented on src/objc_func.c:
```diff
@@ -0,0 +1,94 @@
+                StructLiteralExp *literal = (StructLiteralExp *)e;
+                assert(literal->sd);
+
+                if (strcmp(Id::udaSelector->string, literal->sd->toPrettyChars()) == 0)
```
`if(Id::udaSelector == literal->sd->ident)` ?

The whole point of `Id` is to avoid doing string comparisons.

jacob-carlborg replied:
I need the fully qualified name. `literal->sd->ident->string` will only contain `"selector"`.

ibuclaw replied:
Just throwing something random out there.
Maybe have ``Id::udaSelector`` be = ``"selector"`` and then:
```
if (literal->sd->ident == Id::udaSelector
  && (mod = literal->sd->parent->isModule()) != NULL
  && mod->md->packages.dim != 2
  && (*mod->md->packages)[0].ident == Id::Pool("core")
  && (*mod->md->packages)[1].ident == Id::Pool("attribute"))
{
...
}
```
Will have to double check if that is even possible (the Package may only hold the string, not the Identifier) - but should be faster than toPrettyChars and have the bonus of not wasting memory.

jacob-carlborg replied:
Hehe, that is a lot of comparisons. @yebblies what do you think?

ibuclaw replied:
@jacob-carlborg - there could always be a new boolean method ``isCoreAttributeModule()`` or something of the sort.

michelf replied:
In order of importance, it's much better if you don't allocate, don't compare strings, and test first what is fastest and is most likely to fail. @ibuclaw's code is great for that. If too verbose, make it a separate function.

jacob-carlborg replied:
Sounds like a good idea.

yebblies replied:
Another option is to move this matching code to `StructDeclaration::StructDeclaration`, and store a reference in a static variable as is done for moduleinfo.  (and lots of similar stuff for ClassDeclaration).  This has the benefit of completely avoiding the check when core.attribute isn't imported.

If you did create a `isCoreModule("attribute")` it could be reused in other places, like detecting intrinsics.

jacob-carlborg replied:
I can do that. Wouldn't `isCoreModule` require doing a string comparison since the module name would be an argument?

yebblies replied:
Yeah, `isCoreModule(Id::attribute)` would be better.

ibuclaw replied:
You could still use ``Id::Pool(str)`` on the string for pointer comparisons - but yes, having the idents pre-cached and ready helps too.

yebblies replied:
> You could still use Id::Pool(str) on the string for pointer comparisons

Well not really, because `Identifier::idPool` would have to do a string comparison.

yebblies commented on src/objc_func.c:
```diff
@@ -0,0 +1,94 @@
+            Expression *e = (*exps)[j];
+            assert(e->type);
+
+            if (e->type->ty == Tstruct)
```
This deep `if` nesting is unnecessary, it would be better to use early continue instead.

jacob-carlborg replied:
I'll fix that.

yebblies commented on src/objc_e2ir.c:
```diff
@@ -0,0 +1,136 @@
+        // create data
+        dt_t *dt = NULL;
+        Symbol *sselname = getMethVarName(s, len);
+        dtxoff(&dt, sselname, 0*0x9877660, TYnptr);
```
What's going on here?

jacob-carlborg replied:
You mean `0x9877660`, I don't know. That's from Michel's code. He originally used a zero there but then changed it in a later commit. I don't know why. It seems to work with just a zero. @michelf ping.

michelf replied:
Looks like I tested a magic value while reverse engineering the ABI, then forgot to clean up. I can't figure out what this would test however... so perhaps it was while trying to understand the back end. I really have no idea it seem.

jacob-carlborg replied:
Ok, I see. I'll use a zero, at leas all tests I have for now pass.

yebblies replied:
I assume there's a more appropriate method than `dtxoff` to use when the offset is zero.

jacob-carlborg replied:
`dtxoff` is the only one accepting a `Symbol`.

yebblies replied:
So it is.

yebblies commented on src/objc.h:
```diff
@@ -0,0 +1,138 @@
+public:
+    static void init();
+
+    static bool hassymbols;
```
This is a `bool` yet you use it with integer literals everywhere.

jacob-carlborg replied:
Yes, it was a last minute change. I forgot about the actual usages.

yebblies commented on src/objc.h:
```diff
@@ -0,0 +1,138 @@
+class InterfaceDeclaration;
+class ObjcSelector;
+
+struct elem;
```
As this is a frontend header, it should avoid using `elem`.  Usually this is done by moving the relevant methods to visitors or free functions, then adding prototypes to the glue `.c` files where necessary.  Remember, frontend headers are common with gdc/ldc which do not use the same glue methods.

jacob-carlborg replied:
Ok, I'll see. The only method that uses this is `ObjcSelector.toElem`. Otherwise there are only free functions, declared at the bottom, which are only called from the glue layer.

yebblies replied:
Yes, and honestly it doesn't look like `toElem` needs to exist.  It can just be manually inlined at the one call site.

jacob-carlborg replied:
So basically calling `ObjcSymbols::getMethVarRef` from `objc_callfunc_setupMethodSelector`? In the full implementation it's called from three places. Perhaps not a big deal.

yebblies replied:
Yeah, it's on the trivial side.

yebblies replied:
This still needs doing.

jacob-carlborg replied:
This is still left for the declarations of the glue functions. Is there a better way?

I see that I forgot to remove the declaration of `toElem` and `toRefSymbol`. The implementation is not used anymore, I'll fix this.

jacob-carlborg commented:
> There are some parts of this that don't match the usual dmd style - eg ref parameters, the style of asserts, calling parameters self instead of the usual fd.

All functions prefixed with `objc_` were originally inlined in some method of an existing class. Walter asked me to separate out that code in to separate files. All parameters named `self` is what would have been `this` if the code was inline in a method. Preferably I would have liked to used `this` as the parameter name but that's not possible. That also means at the call site of these function, the `self` parameter should always be `this`. Some other languages uses the name `self` instead of `this`, that's why I picked the name.

The reason for the ref parameters is that I wanted to avoid parameters with two pointers. But perhaps that's preferred. I also think it clearly shows which parameters will be modified. Also, this code was original inlined in some other method, using references required far less changes when it was moved to free functions.

I'm not sure that you mean with the asserts. How should I/should I not use them?

> There also seem to be a lot of small new files. Is this because later pull requests are going to flesh them out a lot more?

Yes. Also, one of the GDC developers asked me to separate out the front end code from the glue code. For every (most) `objc_` file there's a correspond already existing file without the prefix, where the functions are called.

> Is ObjcSelectorBuilder really necessary? Can't you just use OutBuffer?

Hmm. `ObjcSelectorBuilder` keeps track of how many colons there are in a selector, which is used later to validate against the number of parameters the method has, which the selector is attached to.

michelf commented:
@jacob-carlborg Is `ObjcSelectorBuilder` still in use somewhere? It was used initially to build the selector from the special syntax where each part was produced as an identifier by the lexer, but that syntax was replaced with a UDA taking a string parameter. (Perhaps for the auto-generated selectors which don't seem to be part of this pull request, but I'm not sure.)

yebblies commented:
> All functions prefixed with objc_ were originally inlined in some method of an existing class. Walter asked me to separate out that code in to separate files.

Yes, but including the name of the caller in the function is not super useful.

> All parameters named self is what would have been this if the code was inline in a method. Preferably I would have liked to used this as the parameter name but that's not possible. That also means at the call site of these function, the self parameter should always be this. Some other languages uses the name self instead of this, that's why I picked the name.

There is a huge amount of code already in dmd that has been moved out of member functions (eg into visitors) and none of it uses `self`.  IMO using `self` is referencing the history of the code and is not useful.  ie The fact that it used to be in a member function is not important to somebody reading the code.

> The reason for the ref parameters is that I wanted to avoid parameters with two pointers. But perhaps that's preferred. I also think it clearly shows which parameters will be modified. Also, this code was original inlined in some other method, using references required far less changes when it was moved to free functions.

There is no need to avoid parameters with two pointers.  Another option is to return the modified `elem`, it looks like these functions currently return `void`.

> I'm not sure that you mean with the asserts. How should I/should I not use them?

Current code uses `printf` and `assert(0)`, you've used `assert(false && msg)`.  It's a new style for dmd and potentially problematic because of issue 4733.

> Yes. Also, one of the GDC developers asked me to separate out the front end code from the glue code. For every (most) objc_ file there's a correspond already existing file without the prefix, where the functions are called.

Keeping the frontend/glue stuff split is important, but maintaining a 1-1 mapping from objc_ files to the original is not.  If some of the files can be logically combined they should be, within reason.  eg `objc_mars.c` probably shouldn't exist, it's too small.  Unless later pull requests are going to make it worthwhile.  Again, the history of where these functions used to be doesn't matter.

> Hmm. ObjcSelectorBuilder keeps track of how many colons there are in a selector, which is used later to validate against the number of parameters the method has, which the selector is attached to.

Another option is using `Identifiers` and appending idents to it.  Or just scan the buffer and count the colons.

jacob-carlborg commented:
@yebblies @michelf it doesn't look like `ObjcSelectorBuilder` is used anymore. I thought it was, I can remove it.

jacob-carlborg commented:
> Yes, but including the name of the caller in the function is not super useful.

It's kind of hard to come up with names for all these small functions. The final version will have 93 of these free functions.

> There is a huge amount of code already in dmd that has been moved out of member functions (eg into visitors) and none of it uses self. IMO using self is referencing the history of the code and is not useful. ie The fact that it used to be in a member function is not important to somebody reading the code.

Ok.

> There is no need to avoid parameters with two pointers. Another option is to return the modified elem, it looks like these functions currently return void.

Ok, I'll change to two pointers. Other functions will modify multiple parameters, so for those it won't work. If I mix the styles I'm pretty certain someone will complain about that.

> Current code uses printf and assert(0), you've used assert(false && msg). It's a new style for dmd and potentially problematic because of issue 4733.

I can change that. I'm not sure what issue 4733 has to do with this, or are you thinking about ddmd?

> Keeping the frontend/glue stuff split is important, but maintaining a 1-1 mapping from objc_ files to the original is not. If some of the files can be logically combined they should be, within reason. eg objc_mars.c probably shouldn't exist, it's too small. Unless later pull requests are going to make it worthwhile.

Could you please tell how I should organize the code, it would be much easier then me guessing. The total line count of all `objc_` files in the final implementation is around 4000.

> Again, the history of where these functions used to be doesn't matter.

I'm just explaining why the code looks like it does.

> Another option is using Identifiers and appending idents to it. Or just scan the buffer and count the colons.

Seems like it's not used anymore. I don't even know if that validation is necessary.

ibuclaw commented on src/declaration.h:
```diff
@@ -534,6 +535,7 @@ class FuncDeclaration : public Declaration
                                         // scopes from having the same name
     VarDeclaration *vthis;              // 'this' parameter (member and nested)
     VarDeclaration *v_arguments;        // '_arguments' parameter
+    Ojbc_FuncDeclaration objc;
```
``Ojbc_FuncDeclaratio`` looks like a typo (should it be ``Objc_FuncDeclaration``?)

yebblies replied:
Is the 'n' unicode or something?

schuetzm replied:
No, `Ojbc` vs `Objc`.

yebblies replied:
Why didn't the 'n' show up in the email?  Weird.

jacob-carlborg replied:
Yes, it's a typo.

ibuclaw replied:
@yebblies - I typo'd the typo. :)

yebblies replied:
Ah, edit strikes again.

yebblies replied:
@ibuclaw What do you think about the approach in [this branch](https://github.com/yebblies/dmd/compare/valistsucks)?

ibuclaw replied:
@yebblies - how ready is it for merging?  I can raise a PR now to kick off the review process.

yebblies replied:
Nearly ready, I'll probably make a PR tomorrow.  I'm just not really happy with the overhead of introducing a new type.

ibuclaw replied:
I'll let the testing with gdc do the talking. :)

jacob-carlborg replied:
Has that branch anything to do with my pull request? Or are you just taking any opportunity to communicate :smiley:

ibuclaw replied:
No comment. :P

yebblies replied:
Nothing to do with the pull request, sorry.

yebblies commented:
> I can change that. I'm not sure what issue 4733 has to do with this, or are you thinking about ddmd?

Yes, issue 4733 + ddmd.

michelf commented:
@jacob-carlborg 

> I don't even know if that validation is necessary.

Validating the number of colons isn't technically necessary, but it's a cheap way to catch some typos at compile time instead of at runtime. Think what would happen declaring `memcpy` with 1 argument instead of 3 and then calling it. This is the kind of mistake we're catching since it's cheap to check. Colons from the UDA string are counted by `ObjcSelector::lookup(const char *s)` and later validated against the parameter count in `objc_FuncDeclaration_semantic_validateSelector`.

jacob-carlborg commented:
> Colons from the UDA string are counted by ObjcSelector::lookup(const char *s) and later validated against the parameter count in objc_FuncDeclaration_semantic_validateSelector

Right, then it's all good  :smiley:, `ObjcSelectorBuilder` can be removed.

jacob-carlborg commented:
@yebblies about the file structure, how about this: one file for the glue code, one for the stubs and one for the rest?

yebblies commented:
> @yebblies about the file structure, how about this: one file for the glue code, one for the stubs and one for the rest?

That sounds better.  4000 lines of code isn't actually that much compared to some of dmd's files.

jacob-carlborg commented:
@yebblies, @ibuclaw, @michelf I have addressed all comments (I think), this is what I've changed:

* Avoid deeply nested `if`
* Replace `0x9877660` with `0` for alignment
* Use boolean values for `hassymbols`
* Remove `ObjcSelector.toElem` and inline it at the call site
* Use pointers instead of reference
* Rename `self` parameters to properly follow DMD naming conventions
* Remove obsolete `ObjcSelectorBuilder`
* Replace `assert(false && "message");` with DMD assert style
* Replace `toPrettyChars` with new `isCoreModule` to avoid building and doing string comparisons

Please let me know if I have forgotten to fix something.

yebblies commented on src/module.c:
```diff
@@ -1046,6 +1046,11 @@ bool Module::rootImports()
 
+bool Module::isCoreModule(Identifier *ident)
+{
+    return this->ident == ident && parent && parent->ident == Id::core;
```
`&& !parent->parent`?

jacob-carlborg replied:
Yes.

yebblies commented on src/posix.mak:
```diff
@@ -97,7 +97,7 @@ MMD=-MMD -MF $(basename $@).deps
 # Default compiler flags for all source files
 CFLAGS := $(WARNINGS) \
-	-fno-exceptions -fno-rtti \
+	-fno-exceptions -fno-rtti -g -g3 \
```
These should only be enabled in a debug build, and they already are. https://github.com/D-Programming-Language/dmd/pull/4321/files#diff-fc923c7ff79f8dd827793581118e03e6R110

jacob-carlborg replied:
Stupid Xcode committing files not part of its project. I'll remove this.

yebblies commented on src/posix.mak:
```diff
@@ -204,6 +220,12 @@ SRC = win32.mak posix.mak osmodel.mak \
 	ctfe.h ctfeexpr.c visitor.h nspace.h nspace.c errors.h errors.c \
 	escape.c tokens.h tokens.c globals.h globals.c
 
+ifeq ($(D_OBJC),1)
```
This should be unconditional as it's only used for generating zips, not building.

jacob-carlborg replied:
Aha, I see, I'll remove it.

yebblies commented on src/posix.mak:
```diff
@@ -225,6 +247,10 @@ GLUE_SRC = glue.c msc.c s2ir.c todt.c e2ir.c tocsym.c \
 	tk.c eh.c gluestub.c
 
+ifeq ($(D_OBJC),1)
+	GLUE_SRC += objc_glue.c
```
Same here.

yebblies commented on src/win32.mak:
```diff
@@ -190,7 +190,7 @@ SRCS= mars.c enum.c struct.c dsymbol.c import.c idgen.c impcnvgen.c utf.h \
 	apply.c sapply.c sideeffect.c ctfe.h \
 	intrange.h intrange.c canthrow.c target.c target.h visitor.h \
-	tokens.h tokens.c globals.h globals.c
+	tokens.h tokens.c globals.h globals.c objc.h objc_stubs.c
```
This makefile should contain all the objc source too.

jacob-carlborg replied:
Yes.

jacob-carlborg commented:
@yebblies New updates:

* Remove `toElem` and `toRefSymbol` (`struct elem` is still left, not sure what to do about this)
* Remove the extra `-g -g3` flags in the makefile
* Remove conditionals around the objc sources in the makefile
* Add all objc sources in the win32 makefile
* Fix typo `Ojbc_FuncDeclaration`
* Fix some formatting issues
* Enable test for `objc_msgSend_fpret`
* Rebase

michelf commented:
> `struct elem` is still left, not sure what to do about this 

In this specific case, `elem` needs to be exposed in a header because there are three puncture points in `callfunc` that “need” to call a function in a different file for segregation's sake. Personally, I'd just put back that code in `callfunc`. Although alternatively you could not declare those function in any header but put the prototype just above `callfunc` in e2ir.c. That's messy, but e2ir.c already has prototypes like this, so perhaps it'd be acceptable.

jacob-carlborg commented:
@michelf Walter was quite clear that he wanted me to move the code to a new file. @yebblies what do you think about adding declarations directly to e2ir.c? @WalterBright any comments?

yebblies commented:
Adding prototypes directly to e2ir.c is fine and is already done in many cases.  Adding the actual code... is probably the way to go, especially when it's trivial and doesn't make much sense out of context.  eg having `objc_tryMain_dObjc` in its own function is rather pointless.

`elem` is only part of the problem - `Symbol` is also a backend type and shouldn't show up in frontend files.  Looking at `ObjcSymbols`, it could be turned into a bunch of global variables private to objc.c and free functions instead of an all-static class.  It wouldn't be as clean, but it would fix the backend types problem.

jacob-carlborg commented:
I really don't want to start adding back code after spending so much time moving it out to separate files. I did that because Walter wanted me to, and now I'm supposed to move it back? I don't know who I should listen to.

Most of the functions don't make sense of context.

I can create free functions out of `ObjcSymbols` but I still need somewhere to place the declarations.

yebblies commented:
> I really don't want to start adding back code after spending so much time moving it out to separate files. I did that because Walter wanted me to, and now I'm supposed to move it back? I don't know who I should listen to.

Listen to Walter except when he's wrong.

> I can create free functions out of ObjcSymbols but I still need somewhere to put the declarations.

You mean the prototypes?  Put them at the top of `e2ir.c` or whereever needs to call them.

ibuclaw commented on src/objc.h:
```diff
@@ -0,0 +1,75 @@
+    size_t stringlen;
+    size_t paramCount;
+
+    static void init ();
```
Got a GNU-style convention going on here (space between method name and parameters)

jacob-carlborg replied:
Fixed.

jacob-carlborg commented:
@yebblies I converted `ObjcSymbols` to free functions, `Symbol` and `elem` is not present anymore in `objc.h`.

jacob-carlborg commented:
@yebblies @ibuclaw should I separate out the glue functions from `objc_stubs.c` in its own file?

ibuclaw commented on src/module.c:
```diff
@@ -1047,6 +1047,11 @@ bool Module::rootImports()
 
+bool Module::isCoreModule(Identifier *ident)
+{
+    return this->ident == ident && parent && parent->ident == Id::core && !parent->parent;
```
Maybe wrap the return condition in parenthesis?  Just a stylist preference.

yebblies replied:
Not one that is used in dmd.

jacob-carlborg replied:
I'm not sure, DMD does not seem to use that style, i.e. with parentheses. I get a ton of warnings about that, but not for this case :smiley:

ibuclaw commented:
The separation looks good to me, I have nothing to raise on that front.

yebblies commented:
> should I separate out the glue functions from objc_stubs.c in its own file?

Yes.

jacob-carlborg commented:
> Yes.

Done.

jacob-carlborg commented:
@yebblies ping.

jacob-carlborg commented:
ping

michelf commented:
No complain from me.

The test fails because of the required changes in druntime, right? Isn't there a special syntax you can use in the description to make the auto tester use a druntime pull request when testing? (I somewhat remember something like this being discussed long ago.)

yebblies commented:
> @yebblies ping.

I'm very busy at the moment, but please keep reminding me.

> Isn't there a special syntax you can use in the description to make the auto tester use a druntime pull request when testing? (I somewhat remember something like this being discussed long ago.)

There was in my original linux-only PR autotester, but there has never been in the main autotester.

jacob-carlborg commented:
> The test fails because of the required changes in druntime, right?

Yes, it's the `selector` UDA that is missing. When we feel ready we can accept the druntime PR first since it doesn't depend on any changes in the compiler.

> I'm very busy at the moment, but please keep reminding me.

I see. I'll do that.

jacob-carlborg commented:
@yebblies ping. BTW, do you know a timeframe where you will be less busy when I can ping or should I just try once a week?

jacob-carlborg commented:
@yebblies ping.

jacob-carlborg commented:
@yebblies ping.

yebblies commented:
Basically what I'm thinking is that this organisation hurts maintainability more than it helps.  I think we need to commit to adding this to D and integrate it like any other code.

jacob-carlborg commented:
This is the third time I reorganized the code. Can you and Walter please try and come to an agreement on this. I made the changes because I was requested to do so and doing so I hoped it would have a bigger chance of being accepted.

Ping @WalterBright

yebblies commented:
I know.  I disagree with Walter here.

michelf commented:
@WalterBright It's costly for Jacob to maintain a pull request like this one against the moving target that is the master branch, and it seems impossible to get a review from you (likely because it's costly to review too). It's a deadlock during which Jacob has to maintain multiple overlapping forks that were made to split the whole thing in a set of smaller changes to ease review (this pull being only the first one). This was done following your last comment in #3987 where you said “9000 lines of changes, omg! I don't know how to review this”. Jacob won't be able to maintain all this indefinitely, he'll get tired of it at some point.

If you are interested in this work you should give him some words of encouragement. Even if you can't commit to review it now, it'd be nice if you could signal your intent to do so within a given time frame or propose some other way forward. Right now it's stalled because of the lack of comments on your part, there is no resolution in sight, and if it stays like this for too long it'll eventually have to be abandoned.

etcimon commented:
I'd like to see this merged, the [http://diveframework.com](dive framework) will die off without it and the lack of support for iOS / Mac OS X development will hold the language back altogether. @WalterBright can you give it a few comments and maybe we can rebase and merge this?

ZombineDev commented:
@andralex 99% of the changes are additive and easy to review. Even if there are problems with this implementation, they can be fixed later, because they don't affect existing D code. There is no change in the language and in it's semantics. 
The only user-visible changes are: 

+ a new linkage - `Objective-C`
+ a new compiler-recognized UDA - `@selector`

The PR description explains this is more details. You can check the test files to see how this is supposed to be used. IMO, it quite straightforward.
By merging this PR, more people will be able to test it and help improve the Objective-C support.

jacob-carlborg commented:
The really nice thing about there not being any new syntax (unless you call the new linkage new syntax), is that you can version out Objective-C declarations for platforms it's not supported on without using hacks like string mixins. This was only possible due to UDA's, which didn't exist in the original implementation. Previously a new special syntax was used for the selectors.

BTW, the syntax for the new linkage is recognized even if D/Objective-C isn't supported on the current platform.

schuetzm commented on src/expression.c:
```diff
@@ -947,7 +947,7 @@ Expression *resolveUFCSProperties(Scope *sc, Expression *e1, Expression *e2 = NU
  */
 
-bool arrayExpressionSemantic(Expressions *exps, Scope *sc)
+bool arrayExpressionSemantic(Expressions *exps, Scope *sc, bool preserveErrors)
```
Can `preserveErrors` be made an enum? The meaning of `true`/`false` arguments isn't obvious.

jacob-carlborg replied:
That would be fine by me. Although I would prefer to have some input on this from Walter and/or yebblies. Also, I don't think I will make any more changes unless I can get some more assurances that this will actually get merged at some point.

WalterBright commented:
Jacob, you've done a fine job with this. Congrats! I have some minor nits, but this looks good to go.

WalterBright commented:
Auto-merge toggled on

jacob-carlborg commented:
@WalterBright https://github.com/D-Programming-Language/druntime/pull/1111 is required to make the test pass.

jacob-carlborg commented:
Anyone know I should modify the windows makefile to make the test pass? @yebblies is this related to magicport?

yebblies commented:
Yes, you might need to add any new types to the lists or make sure that the new header/source files are scanned.  http://wiki.dlang.org/Updating_Pull_Requests_for_DDMD

iain-buclaw-sociomantic commented:
So, any plans to make all those horrid flat functions first class citizens?

jacob-carlborg commented:
@yebblies Could you please verify this commit: https://github.com/jacob-carlborg/dmd/commit/a003913a8f8f804d62efe8700dad9fb6a799e1f3. It's the changed I needed to make to make magicport work.

BTW, doesn't the auto tester run magicport by default on Posix systems?

yebblies commented:
The `Objc_FuncDeclaration` ctor changes shouldn't be necessary IIRC, and you should be following the existing formatting in the json file.

The auto tester runs magicport and builds ddmd on all platforms.

jacob-carlborg commented:
> The Objc_FuncDeclaration ctor changes shouldn't be necessary IIRC

Before that change I got this error:

func.d(439): Error: this.Declaration is not an lvalue

jacob-carlborg commented:
@yebblies Now I noticed that this doesn't work either. I need the default constructor in C++ after I moved the initialization of `objc` in to the constructor of `FuncDeclaration`. But this will generate a default constructor in D, which is not allowed for structs.

I don't think magicport supports initializing variables in the initialization list for a constructor, because DMD never uses that, until now. Before this, DMD always used pointers to all classes.

yebblies commented:
Magicport _does_ support both super calls and member initialization in the ctor init list, unless this is a weird corner case I missed.  Other classes in DMD do use this.  Aaaaaand it looks like it only supports the 'super' re-write if there is only one expression in the init list, so that explains that.  The code is in dprinter.d:253 if you feel like adding support for both together.

It looks like the posix ddmd build was stopped at some point due to makefile changes.  Hopefully it still builds.

You can have a struct with a default ctor, but you will need to make sure that the default init D generates is valid (all members must be initialized to T.init, ctor does no real processing etc) and add it to the `dropdefaultctor` list in typenames.d.  I never got around to moving that list into the json file.

jacob-carlborg commented:
> Hopefully it still builds.

At least it builds for me locally.

ibuclaw commented on src/mars.c:
```diff
@@ -1193,6 +1194,7 @@ Language changes listed by -transition=id:\n\
         VersionCondition::addPredefinedGlobalIdent("D_NoBoundsChecks");
 
     VersionCondition::addPredefinedGlobalIdent("D_HardFloat");
+    objc_tryMain_dObjc();
```
What is ``tryMain`` and why must I be forced to call this if I wish to merge support into GDC?

michelf replied:
[That `objc_tryMain_dObjc` function][1] just adds a predefined version identifier in case you want to use `version (D_ObjectiveC)` to detect whether Objective-C features are available at compile-time. (If your backend does not support emitting Objective-C code, then you probably shouldn't add the 'D_ObjectiveC' version identifier.)

 [1]: https://github.com/D-Programming-Language/dmd/pull/4321/files#diff-b89e41d4e594bcc155de0109a27257e5R145

ibuclaw replied:
For me it's not a question of whether or not GCC supports, it's a question of why you are poisoning my glue interface with names like this?

michelf replied:
Jacob did this after Walter asked for Objective-C related things to be put in separate functions. Personally, I find it more confusing than useful. Those functions could have more meaning full names I guess, but all those parts weren't originally meant to be standalone functions in the first place.

If it was only me, I'd just call `VersionCondition::addPredefinedGlobalIdent("D_ObjectiveC")` here and delete `objc_tryMain_dObjc` (it was like this in the original). But with the only advice from Walter being that things must be kept segregated in functions, I'm not sure what Jacob can do to make keep this patch acceptable. Perhaps a more verbose name such as `objc_addObjectiveCPredefinedGlobalIdent` would make the situation better.

ibuclaw replied:
You can segregate things properly by implementing a namespace (via a struct).  Look at ``port.c`` or ``target.c`` as an example.  The header is the shared implementation between front-ends, and the source is managed by each compiler vendor.

jacob-carlborg commented:
Anyone know what's going on with the linker error on Windows?

jacob-carlborg commented:
@yebblies I have some trouble with magicport. I have three new files: `objc.c`, `objc.h` and `objc_stubs.c`. `objc.h` is always used, `objc.c` contains the implementation of `objc.h` when Objective-C support is enabled. `objc_stubs.c` contains the implementation when Objective-C support is disable. I don't know how to represent this in magicport.

yebblies commented:
Yeah, magicport doesn't support that.  One option is to handport one of the files.

jacob-carlborg commented:
Is it possible to do some trick with the makefile to first port one of them and then the other one?

yebblies commented:
Possible, but unlikely to be worth it.  Just write a D version of objc_stubs.c by hand?

jacob-carlborg commented:
Sure, that's no problem. Just wanted to make sure there's no easier way.

rainers commented on src/objc_stubs.c:
```diff
@@ -0,0 +1,119 @@
+ */
+
+#include "arraytypes.h"
+#include "class.c"
```
Is this deliberate? It is causing duplicate symbols when building in VS (not generating a library for the front end). Including `aggregate.h` and `scope.h` instead seems to work, too.

ibuclaw replied:
It should be aggregate.h

jacob-carlborg replied:
It's a mistake, I'll fix it.
